src/rdkafka_msgset_reader.c

```diff
@@ -692,8 +741,123 @@ rd_kafka_msgset_reader_msg_v2 (rd_kafka_msgset_reader_t *msetr) {
+                                rktp->rktp_partition,
+                                hdr.Offset, ctrl_data.KeySize);
+
+                rd_kafka_buf_read_i16(rkbuf, &ctrl_data.Type);
+
+                /* Client is uninterested in value of commit marker */
+                rd_kafka_buf_skip(rkbuf, (int32_t)(message_end 
+                        - rd_slice_offset(&rkbuf->rkbuf_reader)));
+
+                switch (ctrl_data.Type) {
+                case RD_KAFKA_CTRL_MSG_COMMIT:
+                        /* always ignore. */
+                        break;
+
+                case RD_KAFKA_CTRL_MSG_ABORT:
+                        if (msetr->msetr_rkb->rkb_rk->rk_conf.isolation_level !=
+                                        RD_KAFKA_READ_COMMITTED)
+                                break;
+
+                        if (unlikely(!msetr->msetr_aborted_txns))
+                                goto unexpected_abort_txn;
+
+                        aborted_txn_start_offset = rd_kafka_aborted_txns_next_aborted_txn_offset_for_pid(
+                                msetr->msetr_aborted_txns, msetr->msetr_v2_hdr->PID);
+
+                        if (unlikely(aborted_txn_start_offset == -1))
+                                goto unexpected_abort_txn;
+
+                        if (unlikely(aborted_txn_start_offset >= hdr.Offset)) {
+                                rd_rkb_log(msetr->msetr_rkb, LOG_ERR, "TXN",
+                                        "%s [%"PRId32"]: "
+                                        "Abort txn ctrl msg bad order "
+                                        "at offset %"PRId64". Expected "
+                                        " before or at %"PRId64,
+                                        rktp->rktp_rkt->rkt_topic->str,
+                                        rktp->rktp_partition,
+                                        hdr.Offset, aborted_txn_start_offset);
+                                break;
+                        }
+
+                        /* This marks the end of this (aborted) transaction,
+                         * advance to next aborted transaction in list */
+                        rd_kafka_aborted_txns_move_to_next(
+                                msetr->msetr_aborted_txns, msetr->msetr_v2_hdr->PID);
+                        break;
+
+unexpected_abort_txn:
+                        rd_rkb_log(msetr->msetr_rkb, LOG_WARNING, "TXN",
+                                "%s [%"PRId32"]: "
+                                "Received abort txn ctrl msg for "
+                                "unknown txn PID %"PRId64" at "
+                                "offset %"PRId64": ignoring",
+                                rktp->rktp_rkt->rkt_topic->str,
+                                rktp->rktp_partition,
+                                msetr->msetr_v2_hdr->PID, hdr.Offset);
+                        break;
+
+                default:
+                        rd_kafka_buf_parse_fail(rkbuf,
```

Comment:
We should simply ignore unknown control types (since they are not for us), not fail.
Should still do a debug log on it though.

Reply:
yep, good point